### PR TITLE
Fix multi-gpu SDXL training

### PR DESCRIPTION
**Pull request by Isotr0py**
- Fix : #994 
- Fix: #997 

- Add `--gradient_as_bucket_view` to reduce VRAM usage in DDP training
- Add `--static_graph` to solve conflict between DDP and gradients checkpoints 
- Freeze the last layer of text_encoder1 in `sdxl_train.py` since it doesn't participate in loss calculation, used to prevent RuntimeError in DDP training

These are related to text_encoders training, so I tested them by training text_encoders on 2 GPUs due to the limited VRAM.

### CODE REVIEW

**Comment by FurkanGozukara** [comment]
are you sure last layer of text_encoder1 is not trained? because i don't want single gpu training to be broken
by the way i did dual gpu training on kaggle and it works i have no error
trained text encoder 1

**Comment by Isotr0py** [comment]
@FurkanGozukara 

SDXL uses the output of the penultimate layer of Text Encoder 1 instead of the last layer. As a result, it won't participate in the loss calculation, but raise a RuntimeError while syncing the gradients for DDP.

The grad of text_encoder1's last layer should be none neither in single GPU nor multi-gpus when I reproduced the RuntimeError.

Can you check the grad of text_encoder1's last layer or the grad between different device? It seems that the grads between GPUs are not synced correctly.

**Comment by FurkanGozukara** [comment]
> @FurkanGozukara
> 
> SDXL uses the output of the penultimate layer of Text Encoder 1 instead of the last layer. As a result, it won't participate in the loss calculation, but raise a RuntimeError while backward the gradients for DDP because it received no grad.
> 
> The grad of text_encoder1's last layer should be none neither in single GPU nor multi-gpus when I reproduced the RuntimeError.
> 
> Can you check the grad of text_encoder1's last layer or the grad between different device? It seems that the grads between GPUs are not synced correctly.

how can I check latest layer difference? i have trained model right now that i can compare

**Comment by Isotr0py** [comment]
You can just add a print after `accelerator.backward(loss)` in training:
```python
print([k for k,v in text_encoder1.named_parameters() if v.grad is None])
```

It should outputs in neither single gpu nor multi-gpus:
```
['text_model.encoder.layers.11.self_attn.k_proj.weight', 'text_model.encoder.layers.11.self_attn.k_proj.bias', 'text_model.encoder.layers.11.self_attn.v_proj.weight', 'text_model.encoder.layers.11.self_attn.v_proj.bias', 'text_model.encoder.layers.11.self_attn.q_proj.weight', 'text_model.encoder.layers.11.self_attn.q_proj.bias', 'text_model.encoder.layers.11.self_attn.out_proj.weight', 'text_model.encoder.layers.11.self_attn.out_proj.bias', 'text_model.encoder.layers.11.layer_norm1.weight', 'text_model.encoder.layers.11.layer_norm1.bias', 'text_model.encoder.layers.11.mlp.fc1.weight', 'text_model.encoder.layers.11.mlp.fc1.bias', 'text_model.encoder.layers.11.mlp.fc2.weight', 'text_model.encoder.layers.11.mlp.fc2.bias', 'text_model.encoder.layers.11.layer_norm2.weight', 'text_model.encoder.layers.11.layer_norm2.bias', 'text_model.final_layer_norm.weight', 'text_model.final_layer_norm.bias']
```

Or you can print to compare between the device:
```python
print(accelerator, THE WEIGHT/GRAD YOU WANT TO COMPARE)
```

**Comment by jihnenglin** [comment]
Thank you very much! I'm still running the SD XL training script, but the output images so far are very promising. Great improvements in details and texture.

**Comment by FurkanGozukara** [comment]
multi gpu broken kaggle training anyone has any guess how to fix?

https://github.com/kohya-ss/sd-scripts/issues/1272
